partiql-lang/src/main/kotlin/org/partiql/lang/eval/Bindings.kt

```diff
@@ -56,6 +56,63 @@ fun PartiqlAst.CaseSensitivity.toBindingCase(): BindingCase = when (this) {
     is PartiqlAst.CaseSensitivity.CaseSensitive -> BindingCase.SENSITIVE
 }
 
+/**
+ * Represents a namespaced global binding.
+ */
+class BindingId(
+    private val qualifier: List<BindingName>,
+    private val id: BindingName
+) : Iterable<BindingName> {
+    /**
+     * For input such as `SELECT a FROM catalog1."schema1".table1`, the identifier is `table1`.
+     */
+    fun getIdentifier(): BindingName = id
+
+    /**
+     * For input such as `SELECT a FROM catalog1."schema1".table1`, the qualifier is `catalog1."schema1"`
+     */
+    fun getQualifier(): List<BindingName> = qualifier
+
+    /**
+     * Returns whether the id is qualified.
+     */
+    fun hasQualifier(): Boolean = qualifier.isNotEmpty()
+
+    /**
+     * Returns a collection of the parts of the identifier.
+     */
+    fun getParts(): List<BindingName> {
+        return qualifier + id
+    }
+
+    override fun iterator(): Iterator<BindingName> {
+        return getParts().iterator()
+    }
+
+    override fun toString(): String {
+        return when (hasQualifier()) {
+            true -> "${qualifier.joinToString(".")}.$id"
+            false -> id.toString()
+        }
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is BindingId) return false
+
+        if (qualifier != other.qualifier) return false
+        if (id != other.id) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = qualifier.hashCode()
+        result = 31 * result + id.hashCode()
+        return result
+    }
```

Comment:
Why can't this be a `data` class to avoid the  `equals`, `hashCode` boilerplate?   AFAICT with a quick glance the only one you really need here is `toString()`.

...

On closer inspection does this class even need custom implementations of `equals` and `hashCode`?  I would consider not implementing those things at all since there are hard questions like what is the correct definition of equality is between instances of `BindingId` and `BindingName`, etc.

Reply:
I've obviously been writing a lot of Java recently 😅